community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -158,29 +351,115 @@ private W combine( WorkUnit<Material,W> batch )
+        // This way, if the successor wakes up before we finish marking the units as done, it will
+        // know that it should wait and complete its task. In fact, it may trail our done-marking
+        // process and unpark threads in parallel.
         while ( batch != stackEnd )
         {
-            batch.done = true;
+            batch.complete();
             batch = batch.next;
         }
     }
 
-    private static class WorkUnit<Material, W extends Work<Material,W>>
+    private static class WorkUnit<Material, W extends Work<Material,W>> extends AtomicInteger
     {
-        final W work;
-        volatile WorkUnit<Material,W> next;
-        volatile boolean done;
+        private static final int STATE_QUEUED = 0b00;
+        private static final int STATE_PARKED = 0b01;
+        private static final int STATE_DONE = 0b10;
+        private static final int STATE_CANCELLED = 0b11;
 
-        private WorkUnit( W work )
+        private final W work;
+        private final WorkUnit<Material,W> stackEnd;
+        private final Thread owner;
+        private volatile WorkUnit<Material,W> next;
+
+        private WorkUnit( W work, Thread owner, WorkUnit<Material,W> stackEnd )
         {
             this.work = work;
+            this.owner = owner;
+            this.stackEnd = stackEnd;
+        }
+
+        void park( long time, TimeUnit unit )
+        {
+            if ( compareAndSet( STATE_QUEUED, STATE_PARKED ) )
+            {
+                LockSupport.parkNanos( unit.toNanos( time ) );
+                compareAndSet( STATE_PARKED, STATE_QUEUED );
+            }
+        }
+
+        void complete()
+        {
+            int previousState = getAndSet( STATE_DONE );
+            if ( previousState == STATE_PARKED )
+            {
+                unpark();
+            }
+        }
+
+        void unpark()
+        {
+            LockSupport.unpark( owner );
+        }
+
+        boolean isDone()
+        {
+            return (get() & STATE_DONE) == STATE_DONE;
+        }
+
+        boolean isCancelled()
+        {
+            return get() == STATE_CANCELLED;
+        }
+
+        boolean cancel()
+        {
+            int state;
+            do
+            {
+                state = get();
+                if ( state == STATE_DONE )
+                {
+                    return false;
+                }
+            }
+            while ( !compareAndSet( state, STATE_CANCELLED ) );
+            if ( state == STATE_PARKED && owner != Thread.currentThread() )
+            {
+                // Side-channel cancel: wake up waiting owner to notify
+                unpark();
+            }
+            return true;
+        }
+
+        void checkCancelled()
+        {
+            if ( isCancelled() )
+            {
+                throw new CancellationException();
+            }
+        }
+    }
+
+    private static final class FutureThrow extends FutureTask<Object>
```

Comment:
I'd have this extend `FutureTask<Void>`, to signal that it doesn't actually return anything.
That would mean that `NULL_CALLABLE` should be declared `Callable<Void>` as well.